test/modules/EdgeHubRestartTester/src/Program.cs

```diff
@@ -0,0 +1,166 @@
+                    restartCount++;
+
+                    // Setup Message Task
+                    Task sendMessageTask = Task.CompletedTask;
+                    if (Settings.Current.MessageEnabled)
+                    {
+                        sendMessageTask =
+                            new MessageEdgeHubConnector(
+                                msgModuleClient,
+                                batchId,
+                                eachTestExpirationTime,
+                                restartTime,
+                                Logger,
+                                cts.Token)
+                            .StartAsync();
+                    }
+
+                    // Setup Direct Method Task
+                    Task directMethodTask = Task.CompletedTask;
+                    if (Settings.Current.DirectMethodEnabled)
+                    {
+                        directMethodTask =
+                            new DirectMethodEdgeHubConnector(
+                                dmModuleClient,
+                                batchId,
+                                eachTestExpirationTime,
+                                restartTime,
+                                restartCount,
+                                Logger,
+                                cts.Token)
+                            .StartAsync();
+                    }
+
+                    // Wait for the two task to be done before do a restart
+                    await Task.WhenAll(new[] { sendMessageTask, directMethodTask });
+
+                    // Wait until the specified restart period to do another restart
+                    await Task.Delay((int)(eachTestExpirationTime - DateTime.UtcNow).TotalMilliseconds, cts.Token);
+                }
+            }
+            catch (Exception e)
+            {
+                Logger.LogError($"Exception caught: {e}");
+                throw;
+            }
+            finally
+            {
+                iotHubServiceClient?.Dispose();
+                dmModuleClient?.Dispose();
+                msgModuleClient?.Dispose();
+            }
+
+            await cts.Token.WhenCanceled();
+            completed.Set();
+            handler.ForEach(h => GC.KeepAlive(h));
+            Logger.LogInformation("EdgeHubRestartTester Main() finished.");
+            return 0;
+        }
+
+        static async Task<Tuple<DateTime, HttpStatusCode>> RestartModules(
+            ServiceClient iotHubServiceClient)
+        {
+            bool isRestartNeeded = true;
+
+            CloudToDeviceMethod c2dMethod = new CloudToDeviceMethod("RestartModule");
+            string payloadSchema = "{{ \"SchemaVersion\": \"1.0\", \"Id\": \"{0}\" }}";
+            string payload = string.Format(payloadSchema, "edgeHub");
+            Logger.LogInformation("RestartModule Method Payload: {0}", payload);
+            c2dMethod.SetPayloadJson(payload);
+
+            while (isRestartNeeded)
```

Comment:
This seems to be always true

Reply:
Yes, I just need a while(1) but want to make it explicit that even in the case of thrown exception.

Reply:
I want to be explicit that it is always true even in the case of an exception. 
Do you prefer just have a `while(true)` instead?

Reply:
We should retry n times (say 3) and throw otherwise.